### PR TITLE
Added Mo's algorithm [C++]

### LINKED ISSUE #28: Mo's Algorithm [CPP]
Explanation link:   http://codeforces.com/blog/entry/43230

### PR DESCRIPTION
**Pull request by sirjan13**
Fixes #28 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

I have added the code according to the structure proposed in the issue

### CODE REVIEW

**Comment by singhpratyush** [comment]
Hi! Thanks for showing interest in contributing to this repository.

Please make sure that you have ticked the points mentioned in PR description correctly.

Thanks again!

**Comment by sirjan13** [comment]
@Monal5031 whats the issue can you explain?

**Comment by Monal5031** [comment]
@sirjan13  It seems you are using space indentation, while this repo is configured to use tab spaces for c and cpp. Use tav indentation throughout the code and travis test will pass.

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+//Problem : http://www.spoj.com/problems/DQUERY/
+//Time Complexity : O((n+q)*sqrt(n))
+
+#include <bits/stdc++.h>
```
Comment (mohitkyadav) [inline Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: Only include required header files.

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+#define ll long long int
+#define mod 1000000007
+#define show(a) for(i=0;i<a.size();i++) cout<<a[i]<<" ";
+#define fi first
+#define se second
```
Comment (mohitkyadav) [inline Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: why not directly use `first` and `second`.

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+#define maxheap priority_queue<int>
+#define minheap priority_queue<int,vector<int>,greater<int>>
+#define mp make_pair
+#define fast_io() cin.sync_with_stdio(false);cout.sync_with_stdio(false);
+#define long_zero 0ll
+#define long_one 1ll
+inline int sbt(int x){return __builtin_popcount(x);}
+using namespace std;
+int freq[1111111];
+int BLOCK;
```
Comment (mohitkyadav) [inline Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: Insert a blank line after `int BLOCK`.

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if(a.se.fi/BLOCK == b.se.fi/BLOCK)
+        return a.se.se>b.se.se;
+    return a.se.fi/BLOCK>b.se.fi/BLOCK;
+}
+int main() {
+    //fast_io()
+    int n;
+    scanf("%d",&n);
+    int a[n+3];
+    for(int i=0;i<n;i++)
```
Comment (mohitkyadav) [inline Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: Please add space around binary operators,
`for(int i = 0; i < n; i++)`

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp (hunk 2 above)
Comment (sirjan13) [reply Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: Actually thats my template for Competitive Programming it saves time.Should i change it?

**Comment by mohitkyadav** [comment]
That's right @singhpratyush sir, We should not change the pattern.

**Comment by sirjan13** [comment]
I am unable to fix the indentation everytime i update it fails the tests :(

File: Square Root Decomposition/MO's Algorithm/MOAlgo.cpp (hunk 2 above)
Comment (mohitkyadav) [reply Square Root Decomposition/MO's Algorithm/MOAlgo.cpp]: @sirjan13 this repository is dedicated to all the learners and beginners, so please keep things straight and easily understandable.

**Comment by singhpratyush** [comment]
@sirjan13: You just need to use tabs over spaces.

**Comment by Monal5031** [comment]
@sirjan13 If you want some help in indenting the code properly you can use `coala` locally, the command to run it locally on `algos` directory as 
```
coala --apply-patches -n
```
but first you need to install coala using 
```
pip3 install coala-bears
```
This will tab indent the code itself.

**Comment by sirjan13** [comment]
Thanks for helping. 👍
Will surely update it by today :)

**Comment by mohitkyadav** [comment]
Closing due to inactivity.

**Pull request closed without merging**